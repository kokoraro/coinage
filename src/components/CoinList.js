import React from 'react';
import Coin from './Coin';

const CoinList = () => {
  return <Coin />
}

export default CoinList;