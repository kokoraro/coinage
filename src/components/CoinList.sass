.coin-list
  list-style: none

  li
    display: grid
    grid-template-columns: repeat(7, 1fr)
    padding: 25px 0
    border-top: 2px solid lightgrey
    position: relative

    &:last-of-type
      border-bottom: 2px solid lightgrey

    &:hover .coin-delete
      display: block

ul
  padding: 0