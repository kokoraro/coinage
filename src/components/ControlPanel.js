import React, { Component } from 'react';
import AddCoinListItem from './AddCoinListItem';

class ControlPanel extends Component {
    constructor(props) {
        super(props);
        this.state = {
            selectCoins: null,
        };
        this.handleAddSubmit = this.handleAddSubmit.bind(this);
    }

    componentDidMount() {
        const { selectCoins } = this.props;
        this.setState({ selectCoins });
    }

    componentWillReceiveProps(nextProps) {
        const { selectCoins } = this.props;
        if (nextProps.selectCoins !== selectCoins) {
            this.setState({ selectCoins });
        }
    }

    handleAddCoinClick(id) {
        const { selectCoins } = this.state;
        const newSelectCoins = selectCoins.map(c => (
            c.id === id ? { ...c, showing: !c.showing } : c
        ));
        this.setState({ selectCoins: newSelectCoins });
    }

    handleAddSubmit() {
        const { selectCoins } = this.state;
        const { handleAddCoins } = this.props;
        const newSelectCoins = selectCoins
            .filter(c => c.showing)
            .map(c => c.id);
        handleAddCoins(newSelectCoins);
    }

    render() {
        const {
            addOpen,
            toggleAddOpen,
            handleCurrency,
            handleRefresh,
        } = this.props;
        const { selectCoins } = this.state;
        const form = addOpen ? (
            <div>
                <h4>choose coins to add...</h4>
                <button
                    type="button"
                    onClick={this.handleAddSubmit}
                    className="add-coins-submit"
                >
                    Submit
                </button>
                <div className="add-coins">
                    {selectCoins.map(c => (
                        <AddCoinListItem
                            handleAddCoinClick={() => this.handleAddCoinClick(c.id)}
                            key={c.id}
                            data={c}
                        />
                    ))}
                </div>
                <button
                    type="button"
                    onClick={this.handleAddSubmit}
                    className="add-coins-submit"
                >
                    Submit
                </button>
            </div>
        ) : null;

        return (
            <div>
                <div className="control-panel">
                    <button
                        type="button"
                        className="add-button"
                        onClick={toggleAddOpen}
                    >
                        Edit
                    </button>
                    <button type="button" onClick={handleCurrency}>Currency</button>
                    <button type="button" onClick={handleRefresh}>Refresh</button>
                </div>
                {form}
            </div>
        );
    }
}

export default ControlPanel;
