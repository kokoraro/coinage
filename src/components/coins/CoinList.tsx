import React from 'react';
import CoinListItem from './CoinListItem';
import LoadingPanel from '../ui/LoadingPanel';
import { useQuery } from 'react-query';
import { Coin, fetchCoinPrice } from '../api/coins';
import { useGlobalStateContext } from '../../utils/GlobalStateProvider';

interface CoinListProps {
    coinData: Coin[];
    loading: boolean;
}

type CoinPriceResponse = { [key: string]: { GBP: string; USD: string } };

const DEFAULT_COIN_PRICE = { GBP: '0', USD: '0' };

const CoinList = ({
    coinData,
    loading,
}: CoinListProps) => {
    const { state } = useGlobalStateContext();
    const { activeCoinCodes } = state;
    const { isLoading: isLoadingPrice, data: priceData } = useQuery<CoinPriceResponse, Error>('coinsPrice', () => fetchCoinPrice(activeCoinCodes), { refetchOnMount: false });

    if (loading || isLoadingPrice) return <LoadingPanel />;

    const favouriteCoins = coinData.filter(coin => activeCoinCodes.includes(coin.code));

    return (
        <div className="coin-list">
            {favouriteCoins.map(coin => (
                <CoinListItem
                    coin={coin}
                    isCurrencyDollar={state.isCurrencyDollar}
                    key={coin.id}
                    coinPrice={priceData?.[coin.code] || DEFAULT_COIN_PRICE}
                />
            ))}
        </div>
    );
}

export default CoinList;
