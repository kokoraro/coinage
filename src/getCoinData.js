import requestData from './requestData';

export function getData(codes) {
  return requestData(codes)
    .then(data => data)
    .catch(error => error);
}