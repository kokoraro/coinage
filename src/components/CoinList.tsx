import React from 'react';
import CoinItem from './Coin';
import '../styles/CoinList.sass';
import { Coin } from '../reducers/coins';

interface CoinList {
    coinData: Coin[];
    currDollar: boolean;
    addOpen: boolean;
    handleDelete: (id: number) => void;
}

const CoinList = ({
    coinData,
    currDollar,
    addOpen,
    handleDelete,
}: CoinList) => (
    <ul className="coin-list">
        {coinData.map(c => (
            <CoinItem
                coinData={c}
                currDollar={currDollar}
                key={c.id}
                addOpen={addOpen}
                handleDelete={handleDelete}
            />
        ))}
    </ul>
);

export default CoinList;
