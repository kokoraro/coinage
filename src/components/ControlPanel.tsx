import React, { Component } from 'react';
import AddCoinListItem from './AddCoinListItem';
import { Coin } from '../redux/coins/types';

interface ControlPanelProps {
    selectCoins: Coin[];
    handleAddCoins: (coins: number[]) => void;
    addOpen: boolean;
    toggleAddOpen: () => void;
    handleCurrency: () => void;
    handleRefresh: () => void;
}

interface ControlPanelState {
    selectCoins: Coin[];
}

class ControlPanel extends Component<ControlPanelProps, ControlPanelState> {
    constructor(props: ControlPanelProps) {
        super(props);
        this.state = {
            selectCoins: [],
        };
        this.handleAddSubmit = this.handleAddSubmit.bind(this);
    }

    componentDidMount() {
        const { selectCoins } = this.props;
        this.setState({ selectCoins });
    }

    componentWillReceiveProps(nextProps: ControlPanelProps) {
        const { selectCoins } = this.props;
        if (nextProps.selectCoins !== selectCoins) {
            this.setState({ selectCoins });
        }
    }

    handleAddCoinClick(id: number) {
        const { selectCoins } = this.state;
        const newSelectCoins = selectCoins.map(c => (
            c.id === id ? { ...c, showing: !c.showing } : c
        ));
        this.setState({ selectCoins: newSelectCoins });
    }

    handleAddSubmit() {
        const { selectCoins } = this.state;
        const { handleAddCoins } = this.props;
        const newSelectCoins = selectCoins
            .filter(c => c.showing)
            .map(c => c.id);
        handleAddCoins(newSelectCoins);
    }

    render() {
        const {
            addOpen,
            toggleAddOpen,
            handleCurrency,
            handleRefresh,
        } = this.props;
        const { selectCoins } = this.state;

        return (
            <div>
                <div className="control-panel">
                    <button
                        type="button"
                        className="add-button"
                        onClick={toggleAddOpen}
                    >
                        Edit
                    </button>
                    <button type="button" onClick={handleCurrency}>Currency</button>
                    <button type="button" onClick={handleRefresh}>Refresh</button>
                </div>
                {addOpen && (
                    <div>
                        <h4>choose coins to add...</h4>
                        <button
                            type="button"
                            onClick={this.handleAddSubmit}
                            className="add-coins-submit"
                        >
                            Submit
                        </button>
                        <div className="add-coins">
                            {selectCoins.map(c => (
                                <AddCoinListItem
                                    handleAddCoinClick={() => this.handleAddCoinClick(c.id)}
                                    key={c.id}
                                    data={c}
                                />
                            ))}
                        </div>
                        <button
                            type="button"
                            onClick={this.handleAddSubmit}
                            className="add-coins-submit"
                        >
                            Submit
                        </button>
                    </div>
                )}
            </div>
        );
    }
}

export default ControlPanel;
