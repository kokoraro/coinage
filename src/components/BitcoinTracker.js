import React, { Component } from 'react';
import CoinList from './CoinList';
import ControlPanel from './ControlPanel';
import Footer from './Footer';
import { getPrice } from '../getCoins';
import { getData } from '../getCoinData';

const FAVOURITES = [
  'BTC', 'XRP', 'LTC', 'ETH', 'XMR', 
  'ZEC', 'DSH', 'GNT', 'ADA', 'XVG',
];

class BitcoinTracker extends Component {
  constructor(props) {
    super(props);
    this.state = {
      coins: [
        { id: 0, code: '...', name: "Loading", price: 0 },
      ],
      currDollar: true,
      addOpen: false,
      coinList: []
    }
    this.updateCoins = this.updateCoins.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
    this.handleAddCoins = this.handleAddCoins.bind(this);
  }

  async updateCoins() {
    const filtered = this.state.coinList.filter(a => a.showing);
    const codes = filtered.map(c => c.code);
    const prices = await getPrice(codes); 
    const coins = filtered.map(c => {
      return {...c, price: prices[c.code] };
    });
    this.setState({coins});
  }

  async getCoinData() {
    const coinData = await getData();
    const coinList = Object.keys(coinData).map((coin, idx) => (
      { 
        id: idx, 
        name: coinData[coin].CoinName,
        imageURL: coinData[coin].ImageUrl,
        code: coin,
        showing: FAVOURITES.includes(coin)
      }
    ));
    this.setState({ coinList });
    this.updateCoins();
  }

  handleAddCoins(ids) {
    const coinList = this.state.coinList.map(c => (
      ids.includes(c.id) ? { ...c, showing: true } : c
    ));
    this.setState({coinList}, () => this.updateCoins());
  }

  handleDelete(id) {
    const coinList = this.state.coinList.map(c => (
      c.id === id ? { ...c, showing: false } : c
    ));
    this.setState({ coinList }, () => this.updateCoins());
  }

  componentDidMount() {
    this.getCoinData();
  }

  render() {
    const coinList = this.state.coinList.filter(c => !c.showing);
    const { coins, currDollar, addOpen } = this.state;

    return (
      <div>
        <div className="container">
          <h1>coinage</h1>
          <CoinList 
            coinData={coins}
            currDollar={currDollar}
            addOpen={addOpen}
            handleDelete={this.handleDelete}
          />
          <ControlPanel 
            selectCoins={coinList}
            handleRefresh={this.updateCoins}
            handleAddCoins={this.handleAddCoins}
            addOpen={addOpen}
            toggleAddOpen={() => this.setState({addOpen: !addOpen})}
            handleCurrency={() => this.setState({currDollar: !currDollar})} 
          />
        </div>
        <Footer />
      </div>
    )
  }
}

export default BitcoinTracker;