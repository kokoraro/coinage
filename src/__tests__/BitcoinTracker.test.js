import React from 'react';
import { shallow } from 'enzyme';
import BitcoinTracker from '../components/BitcoinTracker';
import CoinList from '../components/CoinList';
import ControlPanel from '../components/ControlPanel';

jest.mock('../request');
import { getPrice } from '../getCoins';
jest.mock('../requestData');
import { getData } from '../getCoinData';

const app = shallow(<BitcoinTracker />);

describe("Bitcoin Tracker app", () => {
  
  it("renders correctly", () => {
    expect(app).toMatchSnapshot();
  });

  it("shows CoinList component", () => {
    expect(app.find(CoinList).length).toBe(1);
  });
  
  it("shows ControlPanel component", () => {
    expect(app.find(ControlPanel).length).toBe(1);
  });

  describe('get price using promises', () => {
    it('should load price', () => {
      expect.assertions(1)

      return Promise.resolve(getPrice('BTC'))
        .then(data => expect(data.GBP).toEqual(6249.91))
    });
  });

  describe("get coin data using promises", () => {
    it("fetches coin id", () => {
      expect.assertions(1)

      return Promise.resolve(getData(['BTC']))
        .then(data => expect(data[0].Id).toEqual("1182"))
    });
  });
  
});