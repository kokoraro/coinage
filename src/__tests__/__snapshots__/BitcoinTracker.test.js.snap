// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bitcoin Tracker app renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BitcoinTracker />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          coinage
        </h1>,
        <CoinList
          coinData={
            Array [
              Object {
                "code": "...",
                "id": 0,
                "name": "Loading",
                "price": 0,
              },
            ]
          }
          currDollar={true}
        />,
        <ControlPanel
          addOpen={false}
          handleAddCoins={[Function]}
          handleCurrency={[Function]}
          handleRefresh={[Function]}
          selectCoins={
            Array [
              Object {
                "code": "DOGE",
                "id": 4,
                "name": "Dogecoin",
                "showing": false,
              },
              Object {
                "code": "XMR",
                "id": 5,
                "name": "Monero",
                "showing": false,
              },
              Object {
                "code": "ZEC",
                "id": 6,
                "name": "Zcash",
                "showing": false,
              },
              Object {
                "code": "DSH",
                "id": 7,
                "name": "Dash",
                "showing": false,
              },
              Object {
                "code": "NEO",
                "id": 8,
                "name": "NEO",
                "showing": false,
              },
              Object {
                "code": "GNT",
                "id": 9,
                "name": "Golem",
                "showing": false,
              },
            ]
          }
          toggleAddOpen={[Function]}
        />,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "coinage",
        },
        "ref": null,
        "rendered": "coinage",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "coinData": Array [
            Object {
              "code": "...",
              "id": 0,
              "name": "Loading",
              "price": 0,
            },
          ],
          "currDollar": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "addOpen": false,
          "handleAddCoins": [Function],
          "handleCurrency": [Function],
          "handleRefresh": [Function],
          "selectCoins": Array [
            Object {
              "code": "DOGE",
              "id": 4,
              "name": "Dogecoin",
              "showing": false,
            },
            Object {
              "code": "XMR",
              "id": 5,
              "name": "Monero",
              "showing": false,
            },
            Object {
              "code": "ZEC",
              "id": 6,
              "name": "Zcash",
              "showing": false,
            },
            Object {
              "code": "DSH",
              "id": 7,
              "name": "Dash",
              "showing": false,
            },
            Object {
              "code": "NEO",
              "id": 8,
              "name": "NEO",
              "showing": false,
            },
            Object {
              "code": "GNT",
              "id": 9,
              "name": "Golem",
              "showing": false,
            },
          ],
          "toggleAddOpen": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            coinage
          </h1>,
          <CoinList
            coinData={
              Array [
                Object {
                  "code": "...",
                  "id": 0,
                  "name": "Loading",
                  "price": 0,
                },
              ]
            }
            currDollar={true}
          />,
          <ControlPanel
            addOpen={false}
            handleAddCoins={[Function]}
            handleCurrency={[Function]}
            handleRefresh={[Function]}
            selectCoins={
              Array [
                Object {
                  "code": "DOGE",
                  "id": 4,
                  "name": "Dogecoin",
                  "showing": false,
                },
                Object {
                  "code": "XMR",
                  "id": 5,
                  "name": "Monero",
                  "showing": false,
                },
                Object {
                  "code": "ZEC",
                  "id": 6,
                  "name": "Zcash",
                  "showing": false,
                },
                Object {
                  "code": "DSH",
                  "id": 7,
                  "name": "Dash",
                  "showing": false,
                },
                Object {
                  "code": "NEO",
                  "id": 8,
                  "name": "NEO",
                  "showing": false,
                },
                Object {
                  "code": "GNT",
                  "id": 9,
                  "name": "Golem",
                  "showing": false,
                },
              ]
            }
            toggleAddOpen={[Function]}
          />,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "coinage",
          },
          "ref": null,
          "rendered": "coinage",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "coinData": Array [
              Object {
                "code": "...",
                "id": 0,
                "name": "Loading",
                "price": 0,
              },
            ],
            "currDollar": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "addOpen": false,
            "handleAddCoins": [Function],
            "handleCurrency": [Function],
            "handleRefresh": [Function],
            "selectCoins": Array [
              Object {
                "code": "DOGE",
                "id": 4,
                "name": "Dogecoin",
                "showing": false,
              },
              Object {
                "code": "XMR",
                "id": 5,
                "name": "Monero",
                "showing": false,
              },
              Object {
                "code": "ZEC",
                "id": 6,
                "name": "Zcash",
                "showing": false,
              },
              Object {
                "code": "DSH",
                "id": 7,
                "name": "Dash",
                "showing": false,
              },
              Object {
                "code": "NEO",
                "id": 8,
                "name": "NEO",
                "showing": false,
              },
              Object {
                "code": "GNT",
                "id": 9,
                "name": "Golem",
                "showing": false,
              },
            ],
            "toggleAddOpen": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
